File: server_go/bookings.go
package main

import (
	"context"
	"log"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func populateBooking(ctx context.Context, booking *Booking) {
	var tour Tour
	err := ToursCollection.FindOne(ctx, bson.M{"_id": booking.Tour}).Decode(&tour)
	if err == nil {
		populateTour(ctx, &tour)
		booking.TourData = &tour
	}
}

func GetBookingsHandler(c *gin.Context) {
	opts := options.Find().SetSort(bson.M{"createdAt": -1})
	ctx := context.Background()

	cursor, err := BookingsCollection.Find(ctx, bson.M{}, opts)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Помилка сервера", "error": err.Error()})
		return
	}
	defer cursor.Close(ctx)

	bookings := []Booking{}
	if err = cursor.All(ctx, &bookings); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Помилка сервера", "error": err.Error()})
		return
	}

	if bookings == nil {
		bookings = []Booking{}
	}

	for i := range bookings {
		populateBooking(ctx, &bookings[i])
	}

	c.JSON(http.StatusOK, bookings)
}

func CreateBookingHandler(c *gin.Context) {
	var body struct {
		Tour           string `json:"tour"`
		CustomerName   string `json:"customerName"`
		CustomerEmail  string `json:"customerEmail"`
		CustomerPhone  string `json:"customerPhone"`
		NumberOfPeople int    `json:"numberOfPeople"`
		Notes          string `json:"notes"`
	}

	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Неправильний формат запиту", "error": err.Error()})
		return
	}

	tourObjID, err := primitive.ObjectIDFromHex(body.Tour)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Неправильний ID туру"})
		return
	}

	ctx := context.Background()
	var tour Tour
	err = ToursCollection.FindOne(ctx, bson.M{"_id": tourObjID}).Decode(&tour)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Тур не знайдено"})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Помилка сервера", "error": err.Error()})
		}
		return
	}

	if tour.AvailableSpots < body.NumberOfPeople {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Недостатньо вільних місць"})
		return
	}

	totalPrice := tour.Price * float64(body.NumberOfPeople)

	booking := Booking{
		ID:             primitive.NewObjectID(),
		Tour:           tourObjID,
		CustomerName:   body.CustomerName,
		CustomerEmail:  body.CustomerEmail,
		CustomerPhone:  body.CustomerPhone,
		NumberOfPeople: body.NumberOfPeople,
		TotalPrice:     totalPrice,
		Status:         "pending",
		Notes:          body.Notes,
		PaymentStatus:  "pending",
		CreatedAt:      time.Now(),
		UpdatedAt:      time.Now(),
	}

	_, err = BookingsCollection.InsertOne(ctx, booking)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Помилка сервера", "error": err.Error()})
		return
	}

	// Update available spots
	newSpots := tour.AvailableSpots - body.NumberOfPeople
	_, err = ToursCollection.UpdateOne(
		ctx,
		bson.M{"_id": tourObjID},
		bson.M{"$set": bson.M{"availableSpots": newSpots}},
	)
	if err != nil {
		log.Printf("⚠️ Failed to update available spots for tour %s: %v", tour.ID.Hex(), err)
	}

	populateBooking(ctx, &booking)
	c.JSON(http.StatusCreated, booking)
}

func UpdateBookingHandler(c *gin.Context) {
	id := c.Param("id")
	objID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Неправильний ID бронювання"})
		return
	}

	var updateData bson.M
	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Неправильний формат запиту", "error": err.Error()})
		return
	}

	delete(updateData, "_id")
	delete(updateData, "createdAt")
	updateData["updatedAt"] = time.Now()

	ctx := context.Background()
	var booking Booking
	err = BookingsCollection.FindOneAndUpdate(
		ctx,
		bson.M{"_id": objID},
		bson.M{"$set": updateData},
		options.FindOneAndUpdate().SetReturnDocument(options.After),
	).Decode(&booking)

	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"message": "Бронювання не знайдено"})
		} else {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Помилка сервера", "error": err.Error()})
		}
		return
	}

	populateBooking(ctx, &booking)
	c.JSON(http.StatusOK, booking)
}

File: server_node_backup/routes/bookings.js
const express = require('express');
const router = express.Router();
const Booking = require('../models/Booking');
const Tour = require('../models/Tour');
const { protect, admin } = require('../middleware/auth');

// @route   GET /api/bookings
// @desc    Get all bookings
// @access  Private/Admin
router.get('/', protect, admin, async (req, res) => {
  try {
    const bookings = await Booking.find()
      .populate('tour')
      .sort({ createdAt: -1 });
    res.json(bookings);
  } catch (error) {
    res.status(500).json({ message: 'Помилка сервера', error: error.message });
  }
});

// @route   POST /api/bookings
// @desc    Create new booking
// @access  Public
router.post('/', async (req, res) => {
  try {
    const { tour, customerName, customerEmail, customerPhone, numberOfPeople, notes } = req.body;

    const tourData = await Tour.findById(tour);
    
    if (!tourData) {
      return res.status(404).json({ message: 'Тур не знайдено' });
    }

    if (tourData.availableSpots < numberOfPeople) {
      return res.status(400).json({ message: 'Недостатньо вільних місць' });
    }

    const totalPrice = tourData.price * numberOfPeople;

    const booking = await Booking.create({
      tour,
      customerName,
      customerEmail,
      customerPhone,
      numberOfPeople,
      totalPrice,
      notes
    });

    // Update available spots
    tourData.availableSpots -= numberOfPeople;
    await tourData.save();

    const populatedBooking = await Booking.findById(booking._id).populate('tour');
    res.status(201).json(populatedBooking);
  } catch (error) {
    res.status(500).json({ message: 'Помилка сервера', error: error.message });
  }
});

// @route   PUT /api/bookings/:id
// @desc    Update booking status
// @access  Private/Admin
router.put('/:id', protect, admin, async (req, res) => {
  try {
    const booking = await Booking.findByIdAndUpdate(
      req.params.id,
      req.body,
      { new: true, runValidators: true }
    ).populate('tour');
    
    if (!booking) {
      return res.status(404).json({ message: 'Бронювання не знайдено' });
    }
    
    res.json(booking);
  } catch (error) {
    res.status(500).json({ message: 'Помилка сервера', error: error.message });
  }
});

module.exports = router;
